mantis-control-plane/mantis-control-plane-core/src/main/java/io/mantisrx/server/core/scheduler/SchedulingConstraints.java

```diff
@@ -20,23 +20,43 @@
 import io.mantisrx.shaded.com.google.common.annotations.VisibleForTesting;
 import io.mantisrx.shaded.com.google.common.collect.ImmutableMap;
 import java.util.Map;
+import java.util.Optional;
 import lombok.AllArgsConstructor;
 import lombok.Value;
 
 /**
- * A class that represents scheduling constraints. These constraints include the resource constraints, and a map of scheduling attributes (e.g. jdkVersion:17).
+ * A class that represents scheduling constraints.
+ * These constraints include the resource constraints for scheduling,
+ * the size name that could match the size label of a Task Executor,
+ * and a map of scheduling attributes (e.g., jdkVersion:17).
  */
 @AllArgsConstructor(staticName = "of")
 @Value
 public class SchedulingConstraints {
-    // Defines the resource constraints for scheduling
+    /**
+     * Defines the resource constraints for scheduling
+     */
     MachineDefinition machineDefinition;
 
-    // Additional attributes for scheduling (ie. jdkVersion:17)
+    /**
+     * Optional field to set a predefined size name. When this field is present, the scheduling system prioritizes
+     * matching this field with the size name of a Task Executor Group during the worker allocation process. If no match is found,
+     * the function falls back to a fitness calculation on machine definition.
+     */
+    Optional<String> sizeName;
```

Comment:
why have this as an explicit property? why not pass this via attribute?

This is the semantics I am thinking: I have this resource requirements but with these attributes and the size can be part of this.

Reply:
I did indeed consider this approach of passing the size as an attribute to preserve the existing structure of the scheduling constraints. Nonetheless, I had a few concerns that led me to opt for a different path.

Firstly, regarding readability: Retaining the size as an explicit property could potentially enhance the clarity and readability of the code. Integrating the size into the attributes might lead to confusion, as it's not immediately obvious that it's a "size" attribute among other "required" scheduling characteristics.

Secondly, I have concerns about "hard requirements" vs size: Scheduling attributes - like the JDK version - are sort of "hard requirements," whereas size is not. Mixing these diverse types of parameters might disrupt the original definition and purpose of the attributes.

Thirdly, in terms of handling key prefixes: To distinguish the size from other attributes, we may have to implement a prefix mechanism for the size. This wouldn't only add unnecessary complexity, but also divert from the simplicity and straightforwardness of keeping all attribute keys consistent.

Ultimately, it's about finding a convention that works, so if you have strong feelings about this, I'm open to making the changes you suggest.

Reply:
Good point. I think eventually, it would be better to modify the JobDefinition to include hard and soft constraints instead of representing them as job tags.